### ISSUE TITLE
UnwindSafe and RefUnwindSafe should be in libcore

### ISSUE THREAD

**Issue by matklad** [rust-lang/rust#64086]
I *think* that would be useful to allow pure no_std libraries to implement those traits without needing the `std` cargo feature. 

cc @gnzlbg, you've asked about this #40628.

**Comment by gnzlbg** [comment]
FWIW I also needed `AssertUnwindSafe` in `#[no_std]` back then, but I don't remember why.

**Comment by ohadravid** [comment]
I run into this when changing an `std::sync::Mutex` to a `parking_lot` one:
```rust
use std::panic::RefUnwindSafe;
use std::sync::Mutex as StdMutex;
use parking_lot::Mutex as ParkingLotMutex;

fn assert_ref_unwind_safe<T: RefUnwindSafe>(t: T) {

}

fn main() {
    assert_ref_unwind_safe(StdMutex::new(0));

    assert_ref_unwind_safe(ParkingLotMutex::new(0));
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=fefe52f7f65cf6e1f5d5bc5a0970790d))

Errors:

```
   Compiling playground v0.0.1 (/playground)
error[E0277]: the type `std::cell::UnsafeCell<{integer}>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
  --> src/main.rs:12:5
   |
5  | fn assert_ref_unwind_safe<T: RefUnwindSafe>(t: T) {
   |    ----------------------    ------------- required by this bound in `assert_ref_unwind_safe`
...
12 |     assert_ref_unwind_safe(ParkingLotMutex::new(0));
   |     ^^^^^^^^^^^^^^^^^^^^^^ `std::cell::UnsafeCell<{integer}>` may contain interior mutability and a reference may not be safely transferrable across a catch_unwind boundary
   |
   = help: within `lock_api::mutex::Mutex<parking_lot::raw_mutex::RawMutex, {integer}>`, the trait `std::panic::RefUnwindSafe` is not implemented for `std::cell::UnsafeCell<{integer}>`
   = note: required because it appears within the type `lock_api::mutex::Mutex<parking_lot::raw_mutex::RawMutex, {integer}>`
...

```

But `parking_lot` is `no_std`, so it can't implement it!

**Comment by steveklabnik** [comment]
Triage: no change

**Comment by nico-abram** [comment]
I believe this was implemented in https://github.com/rust-lang/rust/pull/84662 and shipped with rustc 1.56, so this can be closed now
